aws-cpp-sdk-core/source/Aws.cpp

```diff
@@ -111,4 +111,45 @@ namespace Aws
         }
 #endif // USE_AWS_MEMORY_MANAGEMENT
     }
+
+    std::atomic<unsigned> APIWrapper::RefCount = 0;
+    std::mutex APIWrapper::InitGuard;
+
+    APIWrapper::APIWrapper(const SDKOptions& options)
+    {
+        if (RefCount.load() == 0u)
```

Comment:
Hi. I think the separated if test outside the mutex makes the logic not thread-safe. I think a fail case is the following. This is on some computer with 8 CPU cores and a multithreading OS capable of simultaneous thread execution + time-slicing.

1. Two threads have previously run the constructor. Therefore, the value of RefCount = 2.
2. Thread 1 runs the destructor. On line 140, it evals to false because the value of RefCount is 2. Therefore, this thread will run the else clause *sometime* in the future.
3. Thread 2 runs the destructor. On line 140, it evals to false because the value of RefCount is 2. Therefore, this thread will run the else clause sometime in the future.
4. Thread 1 runs line 152 and decrements RefCount. Its value is now 1.
5. Thread 2 runs line 152 and decrements RefCount. Its value is now 0.

This results in no thread calling Aws::ShutdownAPI(). You can also run the above in parallel by steps 2 and 3 running at exactly the same time. And steps 4 and 5 running at the same time. The value in RefCount will not be corrupted because it is atomic. However, the if-test was not atomic/protected and therefore fails these scenarios.

So what if we moved to use only one if test and the atomic.fetch_add/sub()? Here is sample code:

```cpp
std::atomic<size_t> APIWrapper::s_count{ 0 };
std::mutex APIWrapper::s_initGuard{};
Aws::SDKOptions APIWrapper::s_awsOptions{};

APIWrapper::APIWrapper() {
    const size_t origCount = s_count.fetch_add(1);
    std::lock_guard<std::mutex> lock(s_initGuard);
    if (0 == origCount) {
        Aws::InitAPI(s_awsOptions);
    }    
}

APIWrapper::~APIWrapper() {
    const size_t origCount = s_count.fetch_sub(1);
    std::lock_guard<std::mutex> lock(s_initGuard);
    if (1 == origCount) {
        Aws::ShutdownAPI(s_awsOptions);
    }
}
```

This will also fail and is not thread safe.

1. Two threads have previously run the constructor. Therefore, the value of RefCount = 2.
2. Thread 1 runs the destructor. origCount value is 2. RefCount=1.
3. Thread 2 runs the destructor. origCount value is 1. RefCount=0.
4. Thread 3 runs the constructor. origCount value is 0. RefCount=1.
5. Thread 2 evals (1 == origCount) as true. Therefore it runs the main clause and calls Shutdown.
6. Thread 3 tries to call AWS apis and crashes.

I believe what is needed is the following:
1. an atomic/protected variable to hold the RefCount
2. for a given thread, an atomic load-compare of RefCount to see if this thread needs to Init/Shutdown
3. when the thread in (2) determines it needs to do an Init/Shutdown, that no other thread will do an Init/Shutdown
4. when the thread in (2) determines it needs to do an Init/Shutdown, that no other thread will change the value of RefCount. Why? Because if RefCount changes, that means the need to run Init/Shutdown might change.

Given these, I believe a mutex is needed around the: increment/decrement, the test, and the Init/Shutdown. I don't think it is possible to protect only a subset of them. Since they all need to be protected, a mutex and lock guard can work well. And using this, it is no longer necessary to use an atomic for RefCount as this protection would be redundant since it is surrounded by a mutex.

I would recommend code similar to this:

```cpp
size_t APIWrapper::s_count{ 0 };
std::mutex APIWrapper::s_initGuard{};
Aws::SDKOptions APIWrapper::s_awsOptions{};

APIWrapper::APIWrapper() {
    std::lock_guard<std::mutex> lock(s_initGuard);
    s_count++;
    if (1 == s_count) {
        Aws::InitAPI(s_awsOptions);
    }    
}

APIWrapper::~APIWrapper() {
    std::lock_guard<std::mutex> lock(s_initGuard);
    s_count--;
    if (0 == s_count) {
        Aws::ShutdownAPI(s_awsOptions);
    }
}
```

Reply:
I don't think s_count needs modulo arithmetic so it shouldn't be unsigned.

Reply:
OK w/ me. I'm neutral on sign/unsigned in this usage. However, your post made me think again on `size_t`. I no longer think `size_t` is stylistically the right type/macro to use. `size_t` should only be used to represent the size of something (usually in bytes).
Personally, I tend to use unsigned to store ref counts. It is probably because Windows COM objects must use unsigned longs for reference counting, so I'm biased that direction.
I would prefer it be a 32 or 64-wide; tending towards the natural width of an integer on a CPU.
Perhaps `long int`?

Reply:
updated my suggested code to use `long int` and clarified two steps in a fail scenario